lib/Word_Lib/Sgn_Abs.thy

```diff
@@ -0,0 +1,131 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory Sgn_Abs
+  imports Most_significant_bit
+begin
+
+section \<open>@{const sgn} and @{const abs} for @{typ "'a word"}\<close>
+
+subsection \<open>Instances\<close>
+
+text \<open>@{const sgn} on words returns -1, 0, or 1.\<close>
+instantiation word :: (len) sgn
+begin
+
+definition sgn_word :: \<open>'a word \<Rightarrow> 'a word\<close> where
+  \<open>sgn w = (if w = 0 then 0 else if 0 <s w then 1 else -1)\<close>
+
+instance ..
+
+end
+
+
+(* Simplification setup for sgn on numerals *)
+
+lemma word_sgn_0[simp]:
+  "sgn 0 = (0::'a::len word)"
+  by (simp add: sgn_word_def)
+
+lemma word_sgn_1[simp]:
+  "sgn 1 = (1::'a::len word)"
+  by (simp add: sgn_word_def)
+
+lemma word_sgn_max_word[simp]:
+  "sgn (- 1) = (-1::'a::len word)"
+  by (clarsimp simp: sgn_word_def word_sless_alt)
+
+lemmas word_sgn_numeral[simp] = sgn_word_def[where w="numeral w" for w]
+
+
+text \<open>@{const abs} on words is the usual definition.\<close>
+instantiation word :: (len) abs
+begin
+
+definition abs_word :: \<open>'a word \<Rightarrow> 'a word\<close> where
+  \<open>abs w = (if w \<le>s 0 then -w else w)\<close>
+
+instance ..
+
+end
+
+
+(* Simplification setup for abs on numerals *)
+
+lemma word_abs_0[simp]:
+  "\<bar>0\<bar> = (0::'a::len word)"
+  by (simp add: abs_word_def)
+
+lemma word_abs_1[simp]:
+  "\<bar>1\<bar> = (1::'a::len word)"
+  by (simp add: abs_word_def)
+
+lemma word_abs_max_word[simp]:
+  "\<bar>- 1\<bar> = (1::'a::len word)"
+  by (clarsimp simp: abs_word_def word_sle_eq)
+
+lemma word_abs_msb:
+  "\<bar>w\<bar> = (if msb w then -w else w)" for w::"'a::len word"
+  by (simp add: abs_word_def msb_word_iff_sless_0 word_sless_eq)
+
+lemmas word_abs_numeral[simp] = word_abs_msb[where w="numeral w" for w]
+
+
+subsection \<open>Properties\<close>
+
+(* Many of these are from linordered_idom, but need <s instead of < and occasionally
```

Comment:
grepping for linordered_idom doesn't produce anything, either I'm blind or it's a possible typo

Reply:
If you grep in Isabelle there should be plenty. It's one of the algebraic type classes.

Reply:
Yes, I think `ag` and repo-in-repo messed up my checking. All good.